File: apps/ship/src/config/crypto.ts
import { createCipheriv, createDecipheriv, randomBytes } from "node:crypto";

const NONCE_SIZE = 12;
const TAG_SIZE = 16;

/** Random 32-byte AES-256 key, base64 encoded. */
export function generateEncryptKey(): string {
  return randomBytes(32).toString("base64");
}

/** AES-256-GCM. Output = base64(nonce ‖ ciphertext ‖ tag) — byte-compatible with nest's Go implementation. */
export function encrypt(keyBase64: string, plaintext: string): string {
  const key = Buffer.from(keyBase64, "base64");
  const nonce = randomBytes(NONCE_SIZE);
  const cipher = createCipheriv("aes-256-gcm", key, nonce);
  const body = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  return Buffer.concat([nonce, body, tag]).toString("base64");
}

export function decrypt(keyBase64: string, encryptedBase64: string): string {
  const key = Buffer.from(keyBase64, "base64");
  const data = Buffer.from(encryptedBase64, "base64");
  if (data.length < NONCE_SIZE + TAG_SIZE) {
    throw new Error("ciphertext too short");
  }
  const nonce = data.subarray(0, NONCE_SIZE);
  const body = data.subarray(NONCE_SIZE, data.length - TAG_SIZE);
  const tag = data.subarray(data.length - TAG_SIZE);
  const decipher = createDecipheriv("aes-256-gcm", key, nonce);
  decipher.setAuthTag(tag);
  try {
    return Buffer.concat([decipher.update(body), decipher.final()]).toString("utf8");
  } catch (e) {
    throw new Error(`decrypt error: ${e instanceof Error ? e.message : String(e)}`);
  }
}

File: apps/nest/config/crypto.go
package config

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"io"
)

// GenerateEncryptKey creates a random 32-byte key encoded as base64.
func GenerateEncryptKey() (string, error) {
	key := make([]byte, 32)
	if _, err := io.ReadFull(rand.Reader, key); err != nil {
		return "", fmt.Errorf("generate key error: %s", err)
	}
	return base64.StdEncoding.EncodeToString(key), nil
}

// Encrypt encrypts plaintext using AES-256-GCM with the given base64-encoded key.
// Returns base64(nonce + ciphertext).
func Encrypt(keyBase64, plaintext string) (string, error) {
	key, err := base64.StdEncoding.DecodeString(keyBase64)
	if err != nil {
		return "", fmt.Errorf("decode key error: %s", err)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("create cipher error: %s", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("create gcm error: %s", err)
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
		return "", fmt.Errorf("generate nonce error: %s", err)
	}

	ciphertext := gcm.Seal(nonce, nonce, []byte(plaintext), nil)
	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

// Decrypt decrypts base64(nonce + ciphertext) using AES-256-GCM with the given base64-encoded key.
func Decrypt(keyBase64, encryptedBase64 string) (string, error) {
	key, err := base64.StdEncoding.DecodeString(keyBase64)
	if err != nil {
		return "", fmt.Errorf("decode key error: %s", err)
	}

	data, err := base64.StdEncoding.DecodeString(encryptedBase64)
	if err != nil {
		return "", fmt.Errorf("decode ciphertext error: %s", err)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("create cipher error: %s", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("create gcm error: %s", err)
	}

	nonceSize := gcm.NonceSize()
	if len(data) < nonceSize {
		return "", fmt.Errorf("ciphertext too short")
	}

	nonce, ciphertext := data[:nonceSize], data[nonceSize:]
	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return "", fmt.Errorf("decrypt error: %s", err)
	}

	return string(plaintext), nil
}
